File: tools/benchmark-harness/scripts/kreuzberg_extract.py
"""Kreuzberg Python extraction wrapper for benchmark harness.

Supports four modes:
- sync: extract_file_sync() - synchronous extraction
- async: extract_file() - asynchronous extraction
- batch: batch_extract_files_sync() - synchronous batch extraction
- server: persistent mode reading paths from stdin
"""

from __future__ import annotations

import asyncio
import json
import platform
import resource
import sys
import time
from typing import Any


def _get_peak_memory_bytes() -> int:
    """Get peak RSS memory in bytes for the current process.

    Uses resource.getrusage(RUSAGE_SELF). On Linux ru_maxrss is in KB,
    on macOS/BSD it is in bytes.
    """
    usage = resource.getrusage(resource.RUSAGE_SELF)
    if platform.system() == "Linux":
        return usage.ru_maxrss * 1024
    return usage.ru_maxrss

from kreuzberg import (
    ExtractionConfig,
    OcrConfig,
    batch_extract_files_sync,
    extract_file,
    extract_file_sync,
)


def _determine_ocr_used(metadata: dict[str, Any], ocr_enabled: bool) -> bool:
    """Determine if OCR was actually used based on extraction result metadata.

    Mirrors the native Rust adapter logic: OCR is used when format_type is "ocr",
    or when format_type is "image" or "pdf" and OCR was enabled in config.
    """
    format_type = (metadata or {}).get("format_type", "")
    if format_type == "ocr":
        return True
    if format_type in ("image", "pdf") and ocr_enabled:
        return True
    return False


def _build_config(ocr_enabled: bool, output_format: str, *, force_ocr: bool = False) -> ExtractionConfig:
    """Build ExtractionConfig with cache off and the requested output format."""
    config = ExtractionConfig(use_cache=False, output_format=output_format)
    if ocr_enabled:
        config.ocr = OcrConfig(backend="tesseract")
    if force_ocr:
        config.force_ocr = True
    return config


def extract_sync(
    file_path: str, ocr_enabled: bool, output_format: str = "markdown", *, force_ocr: bool = False
) -> dict[str, Any]:
    """Extract using synchronous API."""
    config = _build_config(ocr_enabled, output_format, force_ocr=force_ocr)

    start = time.perf_counter()
    result = extract_file_sync(file_path, config=config)
    duration_ms = (time.perf_counter() - start) * 1000.0

    metadata = result.metadata or {}
    return {
        "content": result.content,
        "metadata": metadata,
        "_extraction_time_ms": duration_ms,
        "_ocr_used": _determine_ocr_used(metadata, ocr_enabled or force_ocr),
        "_peak_memory_bytes": _get_peak_memory_bytes(),
    }


async def extract_async(
    file_path: str, ocr_enabled: bool, output_format: str = "markdown", *, force_ocr: bool = False
) -> dict[str, Any]:
    """Extract using asynchronous API."""
    config = _build_config(ocr_enabled, output_format, force_ocr=force_ocr)

    start = time.perf_counter()
    result = await extract_file(file_path, config=config)
    duration_ms = (time.perf_counter() - start) * 1000.0

    metadata = result.metadata or {}
    return {
        "content": result.content,
        "metadata": metadata,
        "_extraction_time_ms": duration_ms,
        "_ocr_used": _determine_ocr_used(metadata, ocr_enabled or force_ocr),
        "_peak_memory_bytes": _get_peak_memory_bytes(),
    }


def extract_batch_sync(file_paths: list[str], ocr_enabled: bool, output_format: str = "markdown") -> list[dict[str, Any]]:
    """Extract multiple files using batch API."""
    config = _build_config(ocr_enabled, output_format)

    start = time.perf_counter()
    results = batch_extract_files_sync(file_paths, config=config)  # type: ignore[arg-type]
    total_duration_ms = (time.perf_counter() - start) * 1000.0

    per_file_duration_ms = total_duration_ms / len(file_paths) if file_paths else 0

    output = []
    for result in results:
        metadata = result.metadata or {}
        output.append({
            "content": result.content,
            "metadata": metadata,
            "_extraction_time_ms": per_file_duration_ms,
            "_batch_total_ms": total_duration_ms,
            "_ocr_used": _determine_ocr_used(metadata, ocr_enabled),
        })
    return output


def _parse_request(line: str) -> tuple[str, bool]:
    """Parse a request line: JSON object with path+force_ocr, or plain file path."""
    stripped = line.strip()
    if stripped.startswith("{"):
        try:
            req = json.loads(stripped)
            return req.get("path", ""), req.get("force_ocr", False)
        except json.JSONDecodeError:
            pass
    return stripped, False


def run_server(ocr_enabled: bool, output_format: str) -> None:
    """Persistent server mode: read paths from stdin, write JSON to stdout."""
    # Signal readiness after Python + FFI initialization
    print("READY", flush=True)
    for line in sys.stdin:
        file_path, force_ocr = _parse_request(line)
        if not file_path:
            continue
        start = time.perf_counter()
        try:
            payload = extract_sync(file_path, ocr_enabled, output_format, force_ocr=force_ocr)
            print(json.dumps(payload), flush=True)
        except Exception as e:
            duration_ms = (time.perf_counter() - start) * 1000.0
            print(json.dumps({"error": str(e), "_extraction_time_ms": duration_ms, "_ocr_used": False}), flush=True)


def main() -> None:
    ocr_enabled = False
    output_format = "markdown"
    args = []
    for arg in sys.argv[1:]:
        if arg == "--ocr":
            ocr_enabled = True
        elif arg == "--no-ocr":
            ocr_enabled = False
        elif arg.startswith("--format="):
            output_format = arg.split("=", 1)[1]
        else:
            args.append(arg)

    if output_format not in ("markdown", "plaintext"):
        print(f"Error: --format must be 'markdown' or 'plaintext'; got '{output_format}'", file=sys.stderr)
        sys.exit(64)

    if len(args) < 1:
        print("Usage: kreuzberg_extract.py [--ocr|--no-ocr] [--format=markdown|plaintext] <mode> <file_path> [additional_files...]", file=sys.stderr)
        print("Modes: sync, async, batch, server", file=sys.stderr)
        sys.exit(1)

    mode = args[0]
    file_paths = args[1:]

    try:
        if mode == "server":
            run_server(ocr_enabled, output_format)

        elif mode == "sync":
            if len(file_paths) != 1:
                print("Error: sync mode requires exactly one file", file=sys.stderr)
                sys.exit(1)
            payload = extract_sync(file_paths[0], ocr_enabled, output_format)
            print(json.dumps(payload), end="")

        elif mode == "async":
            if len(file_paths) != 1:
                print("Error: async mode requires exactly one file", file=sys.stderr)
                sys.exit(1)
            payload = asyncio.run(extract_async(file_paths[0], ocr_enabled, output_format))
            print(json.dumps(payload), end="")

        elif mode == "batch":
            if len(file_paths) < 1:
                print("Error: batch mode requires at least one file", file=sys.stderr)
                sys.exit(1)

            if len(file_paths) == 1:
                results = extract_batch_sync(file_paths, ocr_enabled, output_format)
                print(json.dumps(results[0]), end="")
            else:
                results = extract_batch_sync(file_paths, ocr_enabled, output_format)
                print(json.dumps(results), end="")

        else:
            print(f"Error: Unknown mode '{mode}'. Use sync, async, batch, or server", file=sys.stderr)
            sys.exit(1)

    except Exception as e:
        print(f"Error extracting with Kreuzberg: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tools/benchmark-harness/scripts/kreuzberg_extract.ts
#!/usr/bin/env tsx
/**
 * Kreuzberg TypeScript/Node.js extraction wrapper for benchmark harness.
 *
 * Supports four modes:
 * - async: extractFile() - asynchronous extraction
 * - batch: batchExtractFiles() - batch extraction for multiple files
 * - async-batch: parallel async extraction
 * - server: persistent mode reading paths from stdin
 */

import * as readline from "node:readline";
import { batchExtractFiles, type ExtractionConfig, extractFile } from "@kreuzberg/node";

interface ExtractionOutput {
	content: string;
	metadata: Record<string, unknown>;
	_extraction_time_ms: number;
	_batch_total_ms?: number;
	_ocr_used: boolean;
	_peak_memory_bytes?: number;
}

/**
 * Determine if OCR was actually used based on extraction result metadata.
 * Mirrors the native Rust adapter logic: OCR is used when format_type is "ocr",
 * or when format_type is "image" and OCR was enabled in config.
 */
function parseRequest(line: string): { path: string; forceOcr: boolean } {
	const trimmed = line.trim();
	if (trimmed.startsWith("{")) {
		try {
			const req = JSON.parse(trimmed);
			return { path: req.path || "", forceOcr: req.force_ocr || false };
		} catch {
			// Fall through to plain path
		}
	}
	return { path: trimmed, forceOcr: false };
}

function determineOcrUsed(metadata: Record<string, unknown>, ocrEnabled: boolean): boolean {
	const formatType = (metadata?.format_type as string) || "";
	if (formatType === "ocr") return true;
	if ((formatType === "image" || formatType === "pdf") && ocrEnabled) return true;
	return false;
}

function createConfig(ocrEnabled: boolean, forceOcr = false): ExtractionConfig {
	return {
		useCache: false,
		...(ocrEnabled && { ocr: { backend: "tesseract", language: "eng" } }),
		...(forceOcr && { forceOcr: true }),
	};
}

async function extractAsync(filePath: string, ocrEnabled: boolean, forceOcr = false): Promise<ExtractionOutput> {
	const config = createConfig(ocrEnabled, forceOcr);
	const start = performance.now();
	const result = await extractFile(filePath, config);
	const durationMs = performance.now() - start;

	const metadata = result.metadata || {};
	return {
		content: result.content,
		metadata,
		_extraction_time_ms: durationMs,
		_ocr_used: determineOcrUsed(metadata as Record<string, unknown>, ocrEnabled || forceOcr),
		_peak_memory_bytes: process.memoryUsage().rss,
	};
}

async function extractBatch(filePaths: string[], ocrEnabled: boolean): Promise<ExtractionOutput[]> {
	const config = createConfig(ocrEnabled);
	const start = performance.now();
	const results = await batchExtractFiles(filePaths, config);
	const totalDurationMs = performance.now() - start;

	const perFileDurationMs = filePaths.length > 0 ? totalDurationMs / filePaths.length : 0;

	const peakMemory = process.memoryUsage().rss;
	return results.map((result) => {
		const metadata = result.metadata || {};
		return {
			content: result.content,
			metadata,
			_extraction_time_ms: perFileDurationMs,
			_batch_total_ms: totalDurationMs,
			_ocr_used: determineOcrUsed(metadata as Record<string, unknown>, ocrEnabled),
			_peak_memory_bytes: peakMemory,
		};
	});
}

async function extractAsyncBatch(filePaths: string[], ocrEnabled: boolean): Promise<ExtractionOutput[]> {
	const start = performance.now();
	const promises = filePaths.map((fp) => extractAsync(fp, ocrEnabled));
	const results = await Promise.all(promises);
	const totalDurationMs = performance.now() - start;

	const peakMemory = process.memoryUsage().rss;
	return results.map((result) => ({
		...result,
		_batch_total_ms: totalDurationMs,
		_peak_memory_bytes: peakMemory,
	}));
}

async function runServer(ocrEnabled: boolean): Promise<void> {
	const rl = readline.createInterface({
		input: process.stdin,
		output: process.stdout,
		terminal: false,
	});

	// Signal readiness after Node + NAPI initialization
	console.log("READY");

	for await (const line of rl) {
		const { path: filePath, forceOcr } = parseRequest(line);
		if (!filePath) {
			continue;
		}
		const start = performance.now();
		try {
			const payload = await extractAsync(filePath, ocrEnabled, forceOcr);
			console.log(JSON.stringify(payload));
		} catch (err) {
			const durationMs = performance.now() - start;
			const error = err as Error;
			console.log(
				JSON.stringify({
					error: error.message,
					_extraction_time_ms: durationMs,
					_ocr_used: false,
					_peak_memory_bytes: process.memoryUsage().rss,
				}),
			);
		}
	}
}

async function main(): Promise<void> {
	let ocrEnabled = false;
	const args: string[] = [];

	for (const arg of process.argv.slice(2)) {
		if (arg === "--ocr") {
			ocrEnabled = true;
		} else if (arg === "--no-ocr") {
			ocrEnabled = false;
		} else {
			args.push(arg);
		}
	}

	if (args.length < 1) {
		console.error("Usage: kreuzberg_extract.ts [--ocr|--no-ocr] <mode> <file_path> [additional_files...]");
		console.error("Modes: async, batch, async-batch, server");
		process.exit(1);
	}

	const mode = args[0];
	const filePaths = args.slice(1);

	try {
		if (mode === "server") {
			await runServer(ocrEnabled);
		} else if (mode === "async") {
			if (filePaths.length !== 1) {
				console.error("Error: async mode requires exactly one file");
				process.exit(1);
			}
			const payload = await extractAsync(filePaths[0], ocrEnabled);
			console.log(JSON.stringify(payload));
		} else if (mode === "batch") {
			if (filePaths.length < 1) {
				console.error("Error: batch mode requires at least one file");
				process.exit(1);
			}

			const results = await extractBatch(filePaths, ocrEnabled);

			if (filePaths.length === 1) {
				console.log(JSON.stringify(results[0]));
			} else {
				console.log(JSON.stringify(results));
			}
		} else if (mode === "async-batch") {
			if (filePaths.length < 1) {
				console.error("Error: async-batch mode requires at least one file");
				process.exit(1);
			}

			const results = await extractAsyncBatch(filePaths, ocrEnabled);

			if (filePaths.length === 1) {
				console.log(JSON.stringify(results[0]));
			} else {
				console.log(JSON.stringify(results));
			}
		} else {
			console.error(`Error: Unknown mode '${mode}'. Use async, batch, async-batch, or server`);
			process.exit(1);
		}
	} catch (err) {
		const error = err as Error;
		console.error(`Error extracting with Kreuzberg: ${error.message}`);
		process.exit(1);
	}
}

main().catch((err) => {
	console.error(err);
	process.exit(1);
});
